segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,407 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import static java.nio.file.attribute.PosixFilePermission.OWNER_WRITE;
+
+/**
+ * Storage adapter for file system based storage.
+ *
+ * Each segment is represented as a single file on the underlying storage.
+ *
+ * Approach to locking:
+ *
+ * This implementation works on the assumption is data written to the storage is always written in append only
+ * fashion. Once a piece of data is written it is never overwritten. Each block of data has an offset assigned to it
+ * and Pravega always tries to write the same data to the same offset.
+ *
+ * With this assumption the only flow when a write call is made to the same offset twice is when ownership of the
+ * segment changes from one host to another and both the hosts are writing to it.
+ *
+ * As write to an offset to a file is idempotent (any attempt to re-write data with the same file offset does not
+ * cause any form of inconsistency), locking is not required.
+ *
+ * In the absence of locking this is the expected behavior in case of ownership change: both the hosts will keep
+ * writing the same data at the same offset till the time the earlier owner gets a notification that it is not the
+ * current owner. Once the earlier owner received this notification, it stops writing to the segment.
+ *
+ */
+@Slf4j
+public class FileSystemStorage implements Storage {
+
+    //region members
+
+    private final FileSystemStorageConfig config;
+    private final ExecutorService executor;
+    private final AtomicBoolean closed;
+
+    //endregion
+
+    //region constructor
+
+    /**
+     * Creates a new instance of the FileSystemStorage class.
+     *
+     * @param config   The configuration to use.
+     * @param executor The executor to use for running async operations.
+     */
+    public FileSystemStorage(FileSystemStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+        this.ensureInitializedAndNotClosed();
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        return supplyAsync( streamSegmentName, () -> syncOpenRead(streamSegmentName));
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
```

Comment:
If we have to break the parameters into multiple lines, it looks nicer to stack them up:

```
SegmentHandle handle, 
long offset, 
byte[] buffer, 
int bufferOffset, 
int length, 
Duration timeout
```